File: app/src/main/java/com/example/veoveo/data/RepositorioPeliculasUsuario.kt
package com.example.veoveo.data

import com.example.veoveo.model.PeliculaUsuario
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withTimeout

// maneja las peliculas del usuario en firebase
class RepositorioPeliculasUsuario {

    private val firestore = FirebaseFirestore.getInstance()
    private val auth = FirebaseAuth.getInstance()

    // devuelve el id del usuario actual
    private fun obtenerIdUsuario(): String? = auth.currentUser?.uid

    // trae las peliculas del usuario segun el estado (por_ver o vista)
    suspend fun obtenerPeliculasPorEstado(estado: String): Result<List<PeliculaUsuario>> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            withTimeout(10000) { // timeout de 10 segundos
                val snapshot = firestore.collection("usuarios")
                    .document(idUsuario)
                    .collection("peliculas")
                    .whereEqualTo("estado", estado)
                    .get()
                    .await()

                val peliculas = snapshot.documents.mapNotNull { it.toObject(PeliculaUsuario::class.java) }
                Result.success(peliculas)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // añade una pelicula a la biblioteca del usuario
    suspend fun agregarPelicula(pelicula: PeliculaUsuario): Result<Unit> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            firestore.collection("usuarios")
                .document(idUsuario)
                .collection("peliculas")
                .document(pelicula.idPelicula.toString())
                .set(pelicula)
                .await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // cambia el estado de una pelicula (de por_ver a vista por ejemplo)
    suspend fun actualizarEstadoPelicula(idPelicula: Int, nuevoEstado: String): Result<Unit> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            firestore.collection("usuarios")
                .document(idUsuario)
                .collection("peliculas")
                .document(idPelicula.toString())
                .update("estado", nuevoEstado)
                .await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // cambia la valoracion en estrellas de una pelicula
    suspend fun actualizarValoracion(idPelicula: Int, valoracion: Int): Result<Unit> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            firestore.collection("usuarios")
                .document(idUsuario)
                .collection("peliculas")
                .document(idPelicula.toString())
                .update("valoracion", valoracion)
                .await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // elimina una pelicula de la biblioteca
    suspend fun eliminarPelicula(idPelicula: Int): Result<Unit> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            firestore.collection("usuarios")
                .document(idUsuario)
                .collection("peliculas")
                .document(idPelicula.toString())
                .delete()
                .await()

            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // comprueba si una pelicula ya esta en la biblioteca
    suspend fun estaPeliculaEnBiblioteca(idPelicula: Int): Result<Boolean> {
        val idUsuario = obtenerIdUsuario() ?: return Result.failure(Exception("Usuario no autenticado"))

        return try {
            val doc = firestore.collection("usuarios")
                .document(idUsuario)
                .collection("peliculas")
                .document(idPelicula.toString())
                .get()
                .await()

            Result.success(doc.exists())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // trae las peliculas de otro usuario segun el estado
    suspend fun obtenerPeliculasPorEstadoDeUsuario(uid: String, estado: String): Result<List<PeliculaUsuario>> {
        return try {
            val snapshot = firestore.collection("usuarios")
                .document(uid)
                .collection("peliculas")
                .whereEqualTo("estado", estado)
                .get()
                .await()

            val peliculas = snapshot.documents.mapNotNull { it.toObject(PeliculaUsuario::class.java) }
            Result.success(peliculas)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: src/services/repositorioPeliculasUsuario.ts
/**
 * ARCHIVO: services/repositorioPeliculasUsuario.ts
 * DESCRIPCIÓN: Gestiona la biblioteca personal de películas de cada usuario.
 * Permite guardar películas como 'por ver' o 'vistas', valorarlas y listarlas.
 * Utiliza subcolecciones dentro del documento de cada usuario en Firestore.
 */

import {
  collection,
  deleteDoc,
  doc,
  getDoc,
  getDocs,
  query,
  setDoc,
  updateDoc,
  where,
} from 'firebase/firestore';
import type { PeliculaUsuario } from '../types';
import { getFirebaseAuth, getFirestoreDb, dbOrThrow, uidOrThrow } from './firebase';

/** Helper: Genera la referencia al documento de una película en la biblioteca del usuario. */
function refPelicula(uid: string, idPelicula: number) {
  return doc(dbOrThrow(), 'usuarios', uid, 'peliculas', String(idPelicula));
}

/** 
 * Recupera la información de una película específica dentro de la biblioteca del usuario actual.
 */
export async function obtenerPeliculaUsuario(idPelicula: number): Promise<PeliculaUsuario | null> {
  try {
    const uid = getFirebaseAuth()?.currentUser?.uid;
    const db = getFirestoreDb();
    if (!uid || !db) return null;
    const snap = await getDoc(doc(db, 'usuarios', uid, 'peliculas', String(idPelicula)));
    if (!snap.exists()) return null;
    return snap.data() as PeliculaUsuario;
  } catch {
    return null;
  }
}

/** 
 * Recupera la información de una película específica de CUALQUIER usuario (ej: un amigo).
 */
export async function obtenerPeliculaDeUsuario(
  uid: string,
  idPelicula: number,
): Promise<PeliculaUsuario | null> {
  const db = getFirestoreDb();
  if (!db) return null;
  const snap = await getDoc(doc(db, 'usuarios', uid, 'peliculas', String(idPelicula)));
  if (!snap.exists()) return null;
  return snap.data() as PeliculaUsuario;
}

/** 
 * Obtiene la lista de películas del usuario actual filtradas por estado (por_ver / vista).
 */
export async function listarPeliculasPorEstado(
  estado: 'por_ver' | 'vista',
): Promise<PeliculaUsuario[]> {
  const uid = uidOrThrow();
  const db = dbOrThrow();
  const q = query(collection(db, 'usuarios', uid, 'peliculas'), where('estado', '==', estado));
  const snap = await getDocs(q);
  return snap.docs.map((d: any) => d.data() as PeliculaUsuario);
}

/** 
 * Obtiene la biblioteca de un usuario específico filtrada por estado.
 */
export async function listarPeliculasPorEstadoDeUsuario(
  uid: string,
  estado: 'por_ver' | 'vista',
): Promise<PeliculaUsuario[]> {
  const db = dbOrThrow();
  const q = query(collection(db, 'usuarios', uid, 'peliculas'), where('estado', '==', estado));
  const snap = await getDocs(q);
  return snap.docs.map((d: any) => d.data() as PeliculaUsuario);
}

/** 
 * Añade una película a la biblioteca del usuario.
 */
export async function agregarPelicula(p: PeliculaUsuario): Promise<void> {
  const uid = uidOrThrow();
  await setDoc(refPelicula(uid, p.idPelicula), p);
}

/** 
 * Cambia el estado de una película (ej: de 'por ver' a 'vista').
 */
export async function actualizarEstadoPelicula(
  idPelicula: number,
  nuevoEstado: 'por_ver' | 'vista',
): Promise<void> {
  const uid = uidOrThrow();
  await updateDoc(refPelicula(uid, idPelicula), { estado: nuevoEstado });
}

/** 
 * Actualiza la puntuación de estrellas (1-10) dada por el usuario.
 */
export async function actualizarValoracion(idPelicula: number, valoracion: number): Promise<void> {
  const uid = uidOrThrow();
  await updateDoc(refPelicula(uid, idPelicula), { valoracion });
}

/** 
 * Elimina una película de la biblioteca del usuario.
 */
export async function eliminarPelicula(idPelicula: number): Promise<void> {
  const uid = uidOrThrow();
  await deleteDoc(refPelicula(uid, idPelicula));
}

/** 
 * Cuenta cuántas películas tiene el usuario en un estado determinado.
 */
export async function contarPeliculasPorEstado(estado: string): Promise<number> {
  try {
    const uid = getFirebaseAuth()?.currentUser?.uid;
    const db = getFirestoreDb();
    if (!uid || !db) return 0;
    const q = query(collection(db, 'usuarios', uid, 'peliculas'), where('estado', '==', estado));
    const snap = await getDocs(q);
    return snap.size;
  } catch {
    return 0;
  }
}

/** 
 * Helper de flujo: Elimina la película y la vuelve a añadir como 'por ver'.
 * Útil para reiniciar el estado de una película rápidamente.
 */
export async function reemplazarPorPorVer(
  idPelicula: number,
  titulo: string,
  rutaPoster: string | null,
): Promise<void> {
  await eliminarPelicula(idPelicula);
  await agregarPelicula({
    idPelicula,
    titulo,
    rutaPoster,
    estado: 'por_ver',
    valoracion: 0,
    fechaAnadido: Date.now(),
  });
}
